src/cli/commands/pushMarketing.ts

```diff
@@ -91,63 +89,51 @@ export class PushMarketing extends Command<ServerContext> {
-      return 1
-    }
-    if (title != null && body == null) {
-      stderr.write('Nothing sent. No body for message.\n')
+    if (body == null || title == null) {
+      stderr.write('Nothing sent. No title or body for message.\n')
       return 1
     }
-    if (body != null && title != null) {
-      const sender = makePushSender(connection)
-      const message = { title, body }
-
-      stdout.write(`Sending push messages...\n`)
-
-      const logger = makeStatusLogger(stdout)
-
-      const payloadMap = new Map<string, Array<Set<string>>>()
-      for await (const deviceRow of streamDevicesByLocation(connection, {
-        country,
-        region,
-        city
-      })) {
-        const { apiKey, deviceId, deviceToken, ignoreMarketing } =
-          deviceRow.device
-
-        // Skip document conditions:
-        if (
-          ignoreMarketing ||
-          apiKey == null ||
-          deviceToken == null ||
-          deviceToken.trim() === ''
-        ) {
-          continue
-        }
-        if (!/^[a-zA-z0-9_\-:]+$/.test(deviceToken)) {
-          // Log invalid tokens
-          logger.write(`Invalid token '${deviceToken}' for doc '${deviceId}'`)
-          continue
-        }
-
-        const payloads = payloadMap.get(apiKey) ?? []
-        const tokens = payloads[payloads.length - 1]
-
-        if (tokens != null && tokens.size < MAX_FIREBASE_BATCH_TOKEN_COUNT) {
-          tokens.add(deviceToken)
-        } else {
-          payloads.push(new Set([deviceToken]))
-        }
-
-        payloadMap.set(apiKey, payloads)
+
+    const sender = makePushSender(connection)
+    const message: SendableMessage = { title, body, isPriceChange: false }
+    const logger = makeStatusLogger(stdout)
+
+    stdout.write('Loading devices...\n')
+
+    // Build a list of all devices we want to send to:
+    const devices = new Map<string, Device>()
+    for await (const deviceRow of streamDevicesByLocation(connection, {
+      country,
+      region,
+      city
+    })) {
+      const { device } = deviceRow
+      const { apiKey, deviceId, deviceToken, ignoreMarketing } = device
+
+      // Skip document conditions:
+      if (
+        ignoreMarketing ||
+        apiKey == null ||
+        deviceToken == null ||
+        deviceToken.trim() === ''
+      ) {
+        continue
       }
 
-      for (const [apiKey, payloads] of payloadMap) {
-        for (const tokens of payloads) {
-          await sender.sendRaw(apiKey, tokens, message).catch(() => {})
-        }
+      if (!/^[a-zA-z0-9_\-:]+$/.test(deviceToken)) {
+        // Log invalid tokens
+        logger.write(`Invalid token '${deviceToken}' for doc '${deviceId}'`)
+        continue
       }
+
+      devices.set(deviceId, device)
+    }
+
+    stdout.write(`Sending to ${devices.size} devices...\n`)
+    for (const device of devices.values()) {
+      try {
+        await sender.sendToDevice(device, message)
+      } catch (error) {}
```

Comment:
Logged to stdout, since we use stderr for the heartbeat. Then you can do `yarn cli push-marketing ... | tee log.txt` to get the interesting output in the file, but not the progress updates.